src/WebJobs.Script/Workers/SharedMemoryDataTransfer/SharedMemoryMap.cs

```diff
@@ -0,0 +1,314 @@
+        /// </summary>
+        private readonly string _mapName;
+
+        public SharedMemoryMap(ILoggerFactory loggerFactory, IMemoryMappedFileAccessor mapAccessor, string mapName, MemoryMappedFile mmf)
+        {
+            if (mmf == null)
+            {
+                throw new ArgumentNullException("mmf");
+            }
+
+            if (string.IsNullOrEmpty(mapName))
+            {
+                throw new ArgumentException("mapName");
+            }
+
+            _loggerFactory = loggerFactory;
+            _logger = _loggerFactory.CreateLogger<SharedMemoryMap>();
+            _mapName = mapName;
+            _mmf = mmf;
+            _mapAccessor = mapAccessor;
+        }
+
+        /// <summary>
+        /// Write content from the given <see cref="Stream"/> into this <see cref="SharedMemoryMap"/>.
+        /// The number of bytes of content being written must be less than or equal to the size of the <see cref="SharedMemoryMap"/>.
+        /// </summary>
+        /// <param name="content"><see cref="Stream"/> containing the content to write.</param>
+        /// <returns>Number of bytes of content written.</returns>
+        public async Task<long> PutStreamAsync(Stream content)
+        {
+            long contentLength = content.Length;
+            byte[] contentLengthBytes = BitConverter.GetBytes(contentLength);
+
+            try
+            {
+                using (MemoryMappedViewStream mmv = _mmf.CreateViewStream())
+                {
+                    await mmv.WriteAsync(contentLengthBytes, 0, SharedMemoryConstants.ContentLengthHeaderBytes);
+                    await content.CopyToAsync(mmv, SharedMemoryConstants.MinBufferSize);
+                    await mmv.FlushAsync();
+                }
+
+                return contentLength;
+            }
+            catch
+            {
+                // This exception can be triggered by:
+                // 1) Size is greater than the total virtual memory.
+                // 2) Access is invalid for the memory-mapped file.
+                _mapAccessor.Delete(_mapName, _mmf);
+            }
+
+            _logger.LogError($"Cannot put bytes: {_mapName}");
+            return 0;
+        }
+
+        /// <summary>
+        /// Write content from the given <see cref="byte[]"/> into this <see cref="SharedMemoryMap"/>.
+        /// The number of bytes of content being written must be less than or equal to the size of the <see cref="SharedMemoryMap"/>.
+        /// TODO gochaudh:
+        /// What happens when byte[] is larger than 2GB? We throw?
+        /// </summary>
+        /// <param name="content"><see cref="byte[]"/> containing the content to write.</param>
+        /// <returns>Number of bytes of content written.</returns>
+        public Task<long> PutBytesAsync(byte[] content)
+        {
+            using (MemoryStream contentStream = new MemoryStream(content))
```

Comment:
Do we know what is the overhead of doing a MemoryStream from an array? Do we duplicate the memory? If so we may want to have a full method for this.

Reply:
Initializing with an array is fine. Does not increase the overhead, it just creates a stream over the given array.
`ToArray()` creates a copy, which I think is fine for now. We can test memory footprint with some sample apps to see what is going on.